_episodes/15.md

```diff
@@ -0,0 +1,356 @@
+
+/**
+ * Write more data from Trino and check data and snapshots
+ */
+INSERT INTO iceberg.logging.logs 
+VALUES 
+(
+  'ERROR', 
+  timestamp '2021-04-01' AT TIME ZONE 'America/Los_Angeles', 
+  'Oh noes', 
+  ARRAY ['Exception in thread "main" java.lang.NullPointerException']
+), 
+(
+  'ERROR', 
+  timestamp '2021-04-01 15:55:23.383345' AT TIME ZONE 'America/Los_Angeles', 
+  'Double oh noes', 
+  ARRAY ['Exception in thread "main" java.lang.NullPointerException']
+), 
+(
+  'WARN', 
+  timestamp '2021-04-01 15:55:23.383345' AT TIME ZONE 'America/Los_Angeles', 
+  'Maybeh oh noes?', 
+  ARRAY ['bad things could be happening']
+);
+
+ 
+SELECT * FROM iceberg.logging.logs;
+
+SELECT * FROM iceberg.logging."logs$snapshots";
+
+/**
+ * Read data from an old snapshot (Time travel)
+ */
+
+SELECT * FROM iceberg.logging."logs@2806470637437034115";
+
+/**
+ * Add new column, notice there is no snapshots of the metadata
+ */
+
+ALTER TABLE iceberg.logging.logs ADD COLUMN severity INTEGER;
+
+SHOW CREATE TABLE iceberg.logging.logs;
+
+SELECT * FROM iceberg.logging.logs;
+
+SELECT * FROM iceberg.logging."logs$snapshots";
+
+/**
+ * Insert new data with new column
+ */
+
+INSERT INTO iceberg.logging.logs VALUES 
+(
+  'INFO', 
+  timestamp '2021-04-01 19:59:59.999999' AT TIME ZONE 'America/Los_Angeles', 
+  'es muy bueno', 
+  ARRAY ['It is all normal'], 
+  1
+);
+
+SELECT * FROM iceberg.logging.logs;
+
+SELECT * FROM iceberg.logging."logs$snapshots";
+
+/**
+ * Rename column and drop column
+ */
+
+ALTER TABLE iceberg.logging.logs RENAME COLUMN severity TO priority;
+
+SHOW CREATE TABLE iceberg.logging.logs;
+
+SELECT * FROM iceberg.logging.logs;
+
+ALTER TABLE iceberg.logging.logs DROP COLUMN priority;
+
+SHOW CREATE TABLE iceberg.logging.logs;
+
+SELECT * FROM iceberg.logging.logs;
+
+SELECT * FROM iceberg.logging."logs$snapshots";
+
+/**
+ * Travel back to previous snapshots
+ */
+
+SELECT * FROM iceberg.logging.logs;
+
+SELECT * FROM iceberg.logging."logs$snapshots";
+
+SELECT * FROM iceberg.logging."logs@<insert-earlier-snapshot>";
+
+CALL system.rollback_to_snapshot('logging', 'logs', <insert-earlier-snapshot>)
+
+/**
+ * Back to the future snapshot
+ */
+
+SELECT * FROM iceberg.logging."logs$snapshots";
+
+SELECT * FROM iceberg.logging."logs@<insert-latest-snapshot>";
+
+CALL system.rollback_to_snapshot('logging', 'logs', <insert-latest-snapshot>)
+
+SELECT * FROM iceberg.logging.logs;
+
+SELECT * FROM iceberg.logging."logs$partitions";
+```
+
+## Question of the week: What do I do to restart the test pipeline if it fails on me?
+
+There is unfortunately no way to enable users to rerun tests without write 
+permissions to the Trino repository, so you have to do a dummy commit.
+
+This can easily be done using this one line hack 
+`git commit --amend --no-edit && git push -f`.
+
+The good news is, we have gone through some extensive lengths to identify flaky
+tests in the last year. This will identify common tests that fail across all of
```

Comment:
in the last year. These test failures are much rarer now, and we are constantly improving the build stability as an ongoing effort.

Reply:
Remove the sentence "This will identify... "